common/lib/plugins/efm2/monitor.ts

```diff
@@ -0,0 +1,305 @@
+
+    this.telemetryFactory.createGauge(`efm2.hostHealthy.${hostId}`, () => (this.hostUnhealthy ? 0 : 1));
+    const task1 = this.newContextRun();
+    const task2 = this.run();
+    Promise.race([task1, task2]).finally(() => {
+      this.stopped = true;
+    });
+  }
+
+  canDispose(): boolean {
+    return this.activeContexts.length === 0 && MonitorImpl.newContexts.size === 0;
+  }
+
+  startMonitoring(context: MonitorConnectionContext): void {
+    if (this.isStopped()) {
+      logger.warn(Messages.get("MonitorImpl.monitorIsStopped", this.hostInfo.host));
+    }
+
+    const startMonitorTimeNano = getCurrentTimeNano() + this.failureDetectionTimeNano;
+    const connectionQueue = MapUtils.computeIfAbsent(
+      MonitorImpl.newContexts,
+      startMonitorTimeNano,
+      () => new Array<WeakRef<MonitorConnectionContext>>()
+    );
+    connectionQueue.push(new WeakRef<MonitorConnectionContext>(context));
+  }
+
+  async newContextRun(): Promise<void> {
+    logger.debug(Messages.get("MonitorImpl.startMonitoringTaskNewContext", this.hostInfo.host));
+
+    try {
+      while (!this.isStopped()) {
+        const currentTimeNanos = getCurrentTimeNano();
+        // Get entries with key (that is a time in nanos) less than current time.
+        const processedKeys: number[] = new Array<number>();
+        for (const [key, val] of MonitorImpl.newContexts.entries()) {
+          if (key < currentTimeNanos) {
+            const queue: Array<WeakRef<MonitorConnectionContext>> = val;
+            processedKeys.push(key);
+            // Each value of found entry is a queue of monitoring contexts awaiting active monitoring.
+            // Add all contexts to an active monitoring contexts queue.
+            // Ignore disposed contexts.
+            let monitorContextRef: WeakRef<MonitorConnectionContext> | undefined;
+            while ((monitorContextRef = queue.shift()) != null) {
+              const monitorContext: MonitorConnectionContext = monitorContextRef.deref();
+              if (monitorContext !== null && monitorContext.isActive()) {
+                this.activeContexts.push(monitorContextRef);
+              }
+            }
+          }
+        }
+        processedKeys.forEach((key) => {
+          MonitorImpl.newContexts.delete(key);
+        });
+        await sleep(1000);
+      }
+      return;
+    } catch (err) {
+      // do nothing, exit task
+    }
+    logger.debug(Messages.get("MonitorImpl.stopMonitoringTaskNewContext", this.hostInfo.host));
+  }
+
+  async run(): Promise<void> {
+    logger.debug(Messages.get("MonitorImpl.startMonitoring", this.hostInfo.host));
+
+    try {
+      while (!this.isStopped()) {
+        try {
+          if (this.activeContexts.length === 0 && !this.hostUnhealthy) {
+            await new Promise((resolve) => {
+              this.delayMillisTimeoutId = setTimeout(resolve, MonitorImpl.TASK_SLEEP_NANOS / 1000000);
```

Comment:
Every time we use `MonitorImpl.TASK_SLEEP_NANOS` we convert it to millis. Does it make sense to just use millis instead? i.e. `MonitorImpl.TASK_SLEEP_MILLIS`